Fixed references encoding in enterprise format

MishaDemianenko opened the pull request:
Variable length encoding of references decoding and encoding or records in enterprise format takes some additional time.
To minimise damage for common case (small records) possibility of fixed records introduced.
From now on in case if record can be encoded in a fixed reference format (criteria based on references that are used in record) it will be used instead of variable-length format.
In case if record can't be encoded in fixed references format - default variable-length format will be selected and used.

changelog:
Speeding up enterprise 'high limit' record format by adding possibility to store small records in a fixed references form to avoid variable length encoding/decoding costs

MishaDemianenko committed:
Fixed references in enterprise format

Because of variable length encoding of references decoding and encoding or records in enterprise format takes some additional time.
To minimize damage for common case of small records possibility of fixed records introduced.
From now on in case if record can be encoded in a fixed reference format (criteria based on references that are used in record) it will be used instead of variable-length format.
In case if record can't be encoded in fixed references format - default variable-length format will be used.

MishaDemianenko committed:
Introduce constants for various bit masks, small refactoring

MishaDemianenko committed:
Mute/remove test output

MishaDemianenko committed:
Add tests for fixed references use cases.
Update node record format to perform correct references evaluations.
Reset record fixed reference flag usage on read.

MishaDemianenko committed:
Update fixed reference record format encoding/decoding and selection criteria.
Introduce unit tests for cases when fixed reference encoding should be used and not used.
Propagate fixed reference flag correctly.

chrisvest commented:
How often are fixed-sized references possible? Looks like the fixed reference size is 32 bit, so this would be used all the time for small stores using the high_limit format?

Will you look at a faster compressed reference format next?

MishaDemianenko committed:
Check if record size is big enough for fixed reference format.

Some formats imply really small record sizes (mostly for testing purposes) that should cause multiple record unit usages.
In some cases it will mean that fixed format can't be used since record size in fixed format will over exceed maximum format record size.

MishaDemianenko committed:
Add description and constant for property format blocks padding bytes

MishaDemianenko committed:
Switch to use inverted fixed reference bit value as marker for fixed reference
to allow new format usage without migration from previous 3.0 format.

Tests for minimal size of fixed references records.

MishaDemianenko committed:
Introduce new high limit format generation for high limit with fixed references.
Introduce 3.0 generation of high limit format.

Add tests for backward compatibility of newly introduced format with fixed references with old format to verify
that we can read records that was stored with base high limit format.

MishaDemianenko committed:
Update java docs.

MishaDemianenko commented:
@chrisvest In case if record can be represented in fixed reference record format then it will be selected and it's correct - most of the cases that will be records in _small_ stores, but also records that do not have big references.

Not in this PR :)

chrisvest commented:
Right, I just noticed that this is a 3.0 change, as opposed to 3.1. Faster compressed references should be a 3.1 thing.

chrisvest commented on enterprise/kernel/src/main/java/org/neo4j/kernel/impl/store/format/highlimit/NodeRecordFormat.java:
```diff
@@ -69,13 +70,32 @@ protected void doReadInternal( NodeRecord record, PageCursor cursor, int recordS
+            long nextProp = decodeCompressedReference( cursor, headerByte, HAS_PROPERTY_BIT, NULL );
+            long labelField = decodeCompressedReference( cursor, headerByte, HAS_LABELS_BIT, NULL_LABELS );
+            record.initialize( inUse, nextProp, dense, nextRel, labelField );
+        }
```
I think it might be good to pull each branch out into their respective method. That would give the JIT more opportunity to not optimise both branches if only one of them is hot. This of course applies to all record formats.

MishaDemianenko replied:
Second one will not be optimised - trap will be there.
I also tried to put those thing out in separate methods to see what effect that will give us - inlining wise and JIT wise it was mostly the same; also there was no perf effects.
The one thing that extraction gives IMHO is readability - its much more easy to see what is the intention of particular branch.
Do you have the same opinion on that ( if so i will happily extract those methods)?

chrisvest replied:
I see that you extract methods for many of these in later commits. I like the readability it gives, but I won't insist on this.

chrisvest commented on enterprise/kernel/src/main/java/org/neo4j/kernel/impl/store/format/highlimit/PropertyRecordFormat.java:
```diff
@@ -68,13 +71,31 @@ public void read( PropertyRecord record, PageCursor cursor, RecordLoad mode, int
-                    toAbsolute( Reference.decode( cursor ), recordId ),
-                    toAbsolute( Reference.decode( cursor ), recordId ) );
+
+            if (record.isUseFixedReferences())
```
This might get complaints from check-style.

chrisvest commented on enterprise/kernel/src/main/java/org/neo4j/kernel/impl/store/format/highlimit/PropertyRecordFormat.java:
```diff
@@ -93,10 +114,31 @@ public void write( PropertyRecord record, PageCursor cursor, int recordSize )
+                cursor.putShort( nextModifier );
+                cursor.putInt( (int) record.getNextProp() );
+                //just stuff 3 bytes
+                cursor.putBytes(new byte[]{0,0,0});
```
You could also just skip those bytes, like the read method does.

chrisvest commented on enterprise/kernel/src/main/java/org/neo4j/kernel/impl/store/format/highlimit/PropertyRecordFormat.java:
```diff
@@ -126,4 +168,10 @@ public long getNextRecordReference( PropertyRecord record )
+    {
+        return (((record.getNextProp() == NULL_REFERENCE.intValue()) || ((record.getNextProp() & 0xFFFF_FFFC_0000_0000L) == 0)) &&
+                ((record.getPrevProp() == NULL_REFERENCE.intValue()) || ((record.getPrevProp() & 0xFFFF_FFFE_0000_0000L) == 0)));
+    }
```
These masks do not match what the read and write methods do.

chrisvest commented on enterprise/kernel/src/main/java/org/neo4j/kernel/impl/store/format/highlimit/NodeRecordFormat.java:
```diff
@@ -48,6 +48,13 @@
+    private static final long RELATIONSHIP_LOWER_BITS_MASK = RELATIONSHIP_HIGHER_BITS_MASK >> 32;
+    private static final long PROPERTY_LOWER_BITS_MASK = PROPERTY_HIGHER_BITS_MASK >> 31;
+    private static final long ONE_BIT_OVERFLOW_BIT_MASK = 0xFFFF_FFFE_0000_0000L;
+    private static final long THREE_BITS_OVERFLOW_BIT_MASK = 0xFFFF_FFFC_0000_0000L;
```
Should this not be `TWO_BITS` instead of `THREE_BITS`?

chrisvest commented on enterprise/kernel/src/main/java/org/neo4j/kernel/impl/store/format/highlimit/NodeRecordFormat.java:
```diff
@@ -114,8 +113,8 @@ protected byte headerBits( NodeRecord record )
-        return (((record.getNextProp() == NULL) || ((record.getNextProp() & THREE_BITS_OVERFLOW_BIT_MASK) == 0)) &&
-                ((record.getNextRel() == NULL) || ((record.getNextRel() & ONE_BIT_OVERFLOW_BIT_MASK) == 0)));
+        return (((record.getNextProp() != NULL) && ((record.getNextProp() & HIGH_DWORD_LOWER_NIBBLE_MASK) == 0)) &&
+                ((record.getNextRel() != NULL) && ((record.getNextRel() & HIGH_DWORD_LOWER_NIBBLE_MASK) == 0)));
```
Why can't NULL references be encoded with fixed references? You are excluding NULL references here, but the fixed-reference write method still tries to handle them.

chrisvest replied:
The read method with its `BaseRecordFormat.longFromIntAndMod` calls should also handle them.

chrisvest commented on enterprise/kernel/src/test/java/org/neo4j/kernel/impl/store/format/highlimit/NodeRecordFormatTest.java:
```diff
@@ -123,6 +123,32 @@ public void useVariableLengthFormatWhenPropertyReferenceTooBig() throws IOExcept
+
+        assertFalse( "Record should use variable length reference if format record is too small.", target.isUseFixedReferences() );
+        verifySameReferences( source, target);
+    }
```
If the record is smaller than what can fit in a fixed-width-reference type record, we use a variable-width-reference type record? Why is that? Fixed-width is supposed to be faster for everything that can fit.

MishaDemianenko replied:
Fixed reference format is a single unit format with fixed number of bytes that required to store particular record data. For example for node record format will require https://github.com/neo4j/neo4j/pull/7670/files#diff-fdb66a08a0f6801727d057d81faa869eR52 bytes at least in a single unit.
In cases when single unit is smaller then that - we should fall back into var length format.

chrisvest replied:
> In cases when single unit is smaller then that - we should fall back into var length format.
> Why? A smaller unit will still fit in a fixed-length unit, right?

MishaDemianenko replied:
Lets take for example `HighLimitWithSmallRecords` record format in this case record size is smaller then required number of bytes, so we will not gonna be able to fit in a single unit

chrisvest replied:
The logic looks inverted to me. Is the `NodeRecordFormat.FIXED_FORMAT_RECORD_SIZE - 1` you give as a `recordSize` parameter the number of bytes that are available for writing the record into, or the number of bytes that the given record is expected to take up?

MishaDemianenko replied:
That is number of bytes available in a format for a single unit.
In this case it's: `NodeRecordFormat.FIXED_FORMAT_RECORD_SIZE - 1` which is 1 byte less then fixed record format requires

chrisvest replied:
Alright, now I get it.

chrisvest commented on enterprise/kernel/src/main/java/org/neo4j/kernel/impl/store/format/highlimit/HighLimit.java:
```diff
@@ -54,7 +54,7 @@
     public HighLimit()
     {
-        super( STORE_VERSION, 7, Capability.DENSE_NODES, Capability.SCHEMA, Capability.LUCENE_5 );
+        super( STORE_VERSION, 8, Capability.DENSE_NODES, Capability.SCHEMA, Capability.LUCENE_5 );
```
Should this not be a separate store version? I also think there should be a `RecordFormats$Factory` implementation for the 3.0 format, with a `META-INF/services` file, so we can load the old implementation.

MishaDemianenko replied:
The reason why factory was not introduced here and store version is not changed - is to preserve compatibility between all 3.0 stores and avoid migration completely. Now we do not need migration at all. do we really want to upgrade format in minor release or we can do it in 3.1 only?

chrisvest replied:
We can't require a migration until 3.1. I just notice that the old format code is still around but not hooked up to anything. I guess you plan to use it to provide migration paths in 3.1?

MishaDemianenko replied:
Correct, the plan is to introduce those things in 3.1

chrisvest commented:
@MishaDemianenko First review done. Looks good. Just one comment about the format versions.

MishaDemianenko committed:
Allow to use fixed reference format for records with null references
